File: scala-example/src/main/scala/demo/actors/SupervisorActor.scala
package demo.actors

import akka.actor.Actor
import akka.actor.ActorLogging
import akka.actor.OneForOneStrategy
import akka.actor.Props
import akka.actor.SupervisorStrategy
import akka.actor.SupervisorStrategy.Escalate
import akka.actor.SupervisorStrategy.Restart
import akka.actor.SupervisorStrategy.Resume
import akka.actor.SupervisorStrategy.Stop

import scala.concurrent.duration.*

/**
 * Supervisor actor that demonstrates proper supervision strategies
 * for fault tolerance and resilience in the actor system.
 */
class SupervisorActor extends Actor with ActorLogging {

  /**
   * Supervision strategy that defines how to handle different types of failures:
   * - IllegalArgumentException: Resume the actor (temporary failure)
   * - NullPointerException: Restart the actor (recoverable failure)
   * - IllegalStateException: Stop the actor (unrecoverable failure)
   * - Exception: Escalate to parent supervisor (unknown failure)
   */
  override val supervisorStrategy: SupervisorStrategy = OneForOneStrategy(
    maxNrOfRetries = 10,
    withinTimeRange = 1.minute,
    loggingEnabled = true
  ) {
    case _: IllegalArgumentException =>
      log.warning("Resuming actor after IllegalArgumentException")
      Resume
    case _: NullPointerException =>
      log.warning("Restarting actor after NullPointerException")
      Restart
    case _: IllegalStateException =>
      log.error("Stopping actor after IllegalStateException")
      Stop
    case _: Exception =>
      log.error("Escalating unknown exception to parent")
      Escalate
  }

  override def preStart(): Unit = {
    log.info("SupervisorActor started")
  }

  override def postStop(): Unit = {
    log.info("SupervisorActor stopped")
  }

  override def receive: Receive = {
    case CreateChild(name, props) =>
      val child = context.actorOf(props, name)
      sender() ! ChildCreated(name, child)
      log.info(s"Created child actor: $name")

    case message =>
      log.info(s"SupervisorActor received: $message")
  }
}

object SupervisorActor {
  def props(): Props = Props(new SupervisorActor)
}

// Messages for supervisor interaction
case class CreateChild(name: String, props: Props)
case class ChildCreated(name: String, ref: akka.actor.ActorRef)

File: java-example/src/main/java/demo/actors/SupervisorActor.java
package demo.actors;

import java.util.concurrent.TimeUnit;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.OneForOneStrategy;
import akka.actor.Props;
import akka.actor.SupervisorStrategy;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import scala.concurrent.duration.Duration;

/**
 * Supervisor actor that demonstrates proper supervision strategies
 * for fault tolerance and resilience in the actor system.
 *
 * Equivalent to Scala's SupervisorActor
 */
public class SupervisorActor extends AbstractActor {

    private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

    /**
     * Supervision strategy that defines how to handle different types of failures:
     * - IllegalArgumentException: Resume the actor (temporary failure)
     * - NullPointerException: Restart the actor (recoverable failure)
     * - IllegalStateException: Stop the actor (unrecoverable failure)
     * - Exception: Escalate to parent supervisor (unknown failure)
     */
    @Override
    public SupervisorStrategy supervisorStrategy() {
        return new OneForOneStrategy(
            10, // maxNrOfRetries
            Duration.create(1, TimeUnit.MINUTES), // withinTimeRange
            akka.japi.pf.DeciderBuilder
                .match(IllegalArgumentException.class, ex -> {
                    log.warning("Resuming actor after IllegalArgumentException");
                    return SupervisorStrategy.resume();
                })
                .match(NullPointerException.class, ex -> {
                    log.warning("Restarting actor after NullPointerException");
                    return SupervisorStrategy.restart();
                })
                .match(IllegalStateException.class, ex -> {
                    log.error("Stopping actor after IllegalStateException");
                    return SupervisorStrategy.stop();
                })
                .match(Exception.class, ex -> {
                    log.error("Escalating unknown exception to parent");
                    return SupervisorStrategy.escalate();
                })
                .build()
        );
    }

    public static Props props() {
        return Props.create(SupervisorActor.class);
    }

    @Override
    public void preStart() {
        log.info("SupervisorActor started");
    }

    @Override
    public void postStop() {
        log.info("SupervisorActor stopped");
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
            .match(CreateChild.class, this::handleCreateChild)
            .matchAny(message -> log.info("SupervisorActor received: {}", message))
            .build();
    }

    private void handleCreateChild(CreateChild message) {
        ActorRef child = getContext().actorOf(message.props(), message.name());
        getSender().tell(new ChildCreated(message.name(), child), getSelf());
        log.info("Created child actor: {}", message.name());
    }

    // Messages for supervisor interaction
    public record CreateChild(String name, Props props) {}
    public record ChildCreated(String name, ActorRef ref) {}
}
